pkg/storegateway/bucket_index_reader.go

```diff
@@ -625,49 +632,74 @@ func (r *bucketIndexReader) preloadSeries(ctx context.Context, ids []storage.Ser
 	return loaded, g.Wait()
 }
 
+var seriesOffsetReaders = &sync.Pool{New: func() any {
+	return &offsetTrackingReader{r: bufio.NewReaderSize(nil, 32*1024)}
+}}
+
+// loadSeries expects the provided ids to be sorted.
 func (r *bucketIndexReader) loadSeries(ctx context.Context, ids []storage.SeriesRef, refetch bool, start, end uint64, loaded *bucketIndexLoadedSeries, stats *safeQueryStats) error {
-	begin := time.Now()
+	defer r.recordLoadSeriesStats(stats, refetch, len(ids), start, end, time.Now())
 
-	b, err := r.block.readIndexRange(ctx, int64(start), int64(end-start))
+	reader, err := r.block.indexRangeReader(ctx, int64(start), int64(end-start))
 	if err != nil {
 		return errors.Wrap(err, "read series range")
 	}
+	defer runutil.CloseWithLogOnErr(r.block.logger, reader, "loadSeries close range reader")
 
-	stats.update(func(stats *queryStats) {
-		stats.seriesFetchCount++
-		stats.seriesFetched += len(ids)
-		stats.seriesFetchDurationSum += time.Since(begin)
-		stats.seriesFetchedSizeSum += int(end - start)
-	})
+	offsetReader := seriesOffsetReaders.Get().(*offsetTrackingReader)
+	defer seriesOffsetReaders.Put(offsetReader)
 
-	for i, id := range ids {
-		c := b[uint64(id)-start:]
+	offsetReader.Reset(start, reader)
+	defer offsetReader.Reset(0, nil) // don't retain the underlying reader
+
+	bytesPool := pool.NewSlabPool[byte](seriesBytesSlicePool, seriesBytesSlabSize)
+	// We never release the pool and let the GC collect it in order to avoid a rance condition with the async sets of the cache.
```

Comment:
[nit] I would explain why we use a pool anyway (to reduce the number of allocations).